### PR TITLE
Remove out-of-support targets from test matrix

### PR DESCRIPTION
**Pull request by abhipsaMisra**
With this new version we can update out test matrix to only test against supported targets. Targets that have reached EOL will no longer be tested against.

Support matrix: https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

Note: .NET 3.1 is also set to reach EOL in Dec 22. Should we remove this target from our testing matrix?

<html>
<body>

Version | Original release date | Latest patch version | Patch release date | Release type | Support phase | End of support
-- | -- | -- | -- | -- | -- | --
.NET Core 3.1 | December 3, 2019 | 3.1.31 | November 8, 2022 | LTS | Maintenance | December 13, 2022

</body>
</html>

### CODE REVIEW

File: e2e/test/E2ETests.csproj [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
     <LangVersion>9.0</LangVersion>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <IsPackable>false</IsPackable>
     <RootNamespace>Microsoft.Azure.Devices.E2ETests</RootNamespace>
     <AssemblyName>Microsoft.Azure.Devices.E2ETests</AssemblyName>
     <RootDir>$(MSBuildProjectDirectory)\..\..</RootDir>
+    <!-- We will turn on this suppression once we are ready to GA-->
     <!-- Suppress framework end of life warnings as we have to keep supporting these frameworks for our customers -->
-    <CheckEolTargetFramework>false</CheckEolTargetFramework>
+    <!-- <CheckEolTargetFramework>false</CheckEolTargetFramework> -->
```
Comment (abhipsaMisra) [inline e2e/test/E2ETests.csproj:12]: We'll add this suppression back once we are ready to GA

File: vsts/vsts.yaml [hunk 2]
```diff
@@ -43,9 +40,7 @@ variables:
   testNetcore31: ${{ or(eq(variables['minMatrix'], 'True'), contains(parameters['testTargets'], 'netcore3.1')) }}
 
 # The remaining build/test targets.
-  testNet50: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'net5.0')) }}
-  testNetcore21: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'netcore2.1')) }}
-  testNet472: ${{ or(eq(variables['allTestTargets'], 'True'), contains(parameters['testTargets'], 'net4.7.2')) }}
+  # Add remaining targets in here
```
Comment (abhipsaMisra) [inline vsts/vsts.yaml:43]: I've left this template in here for future reference

File: e2e/test/E2ETests.csproj (hunk 1 above)
Comment (drwill-ms) [reply e2e/test/E2ETests.csproj:12]: Why not turn it on now?

Comment (drwill-ms) [reply e2e/test/E2ETests.csproj:12]: Oh, in case something else goes out of support, and we want to change. Got it.

File: vsts/vsts.yaml (hunk 2 above)
Comment (drwill-ms) [reply vsts/vsts.yaml:43]: Shall we add .net 7.0 testing?

**Review by drwill-ms — APPROVED** [review]

**Review by tmahmood-microsoft — APPROVED** [review]

**Comment by drwill-ms** [comment]
Need to update the supported platforms file, and probably the migraiton guide.

File: vsts/vsts.yaml (hunk 2 above)
Comment (abhipsaMisra) [reply vsts/vsts.yaml:43]: How about we remove net3.1 and add net 7 (net 3.1 is set to go out of support end of this year).

**Review by schoims — APPROVED** [review]

File: vsts/vsts.yaml (hunk 2 above)
Comment (drwill-ms) [reply vsts/vsts.yaml:43]: k

File: vsts/vsts.yaml [hunk 3]
```diff
@@ -101,13 +91,18 @@ jobs:
             Write-Host "minMatrix: ${{ variables.minMatrix }}"
             Write-Host "testNet60: ${{ variables.testNet60 }}"
             Write-Host "testNetcore31: ${{ variables.testNetcore31 }}"
```
Comment (andyk-ms) [inline vsts/vsts.yaml:93]: Do we need to remove this?

File: vsts/vsts.yaml [hunk 4]
```diff
@@ -398,50 +357,25 @@ jobs:
...
```
Comment (andyk-ms) [inline vsts/vsts.yaml:359]: Do we need to remove this?

**Review by andyk-ms — APPROVED** [review]
Double check on the variable for 3.1 which we are removing.

File: vsts/vsts.yaml (hunk 4 above)
Comment (abhipsaMisra) [reply vsts/vsts.yaml:382]: Retaining these logging statements

File: supported_platforms.md [hunk 5]
```diff
@@ -20,6 +20,8 @@ Nightly test platform details:
 .NET versions tested on
 - .NET 7.0
 - .NET 6.0
+- .NET Core 3.1
+- .NET Framework 4.6.1 (only IoT Hub SDKs)
```
Comment (drwill-ms) [inline supported_platforms.md:24]: Why only Hub SDKs? Is net461 not .netstandard2.0 compliant?

**Review by drwill-ms — APPROVED** [review]

File: supported_platforms.md (hunk 5 above)
Comment (abhipsaMisra) [reply supported_platforms.md]: net461 isn't fully compliant with ns2.0. For this reason I think we should test against net472 instead (which is what we have been doing). What do you think? @drwill-ms 

From https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0#select-net-standard-version:
The versions listed here represent the rules that NuGet uses to determine whether a given .NET Standard library is applicable. While NuGet considers .NET Framework 4.6.1 as supporting .NET Standard 1.5 through 2.0, there are several issues with consuming .NET Standard libraries that were built for those versions from .NET Framework 4.6.1 projects. For .NET Framework projects that need to use such libraries, we recommend that you upgrade the project to target .NET Framework 4.7.2 or higher.

Comment (drwill-ms) [reply supported_platforms.md]: Wow, I wonder why track2 tests with it then.

I'd say stick with 472 for now, but ask the azure sdk team about it.

**Pull request merged**